test/org/sagebionetworks/bridge/validators/StudyParticipantValidatorTest.java

```diff
@@ -94,122 +102,186 @@ public void validatesUpdate() {
-        validator = new StudyParticipantValidator(study, true);
+        validator = new StudyParticipantValidator(externalIdService, study, true);
         assertValidatorMessage(validator, withPhone("234567890", "US"), "phone", "does not appear to be a phone number");
     }
     
     @Test
     public void validatePhone() {
-        validator = new StudyParticipantValidator(study, true);
+        validator = new StudyParticipantValidator(externalIdService, study, true);
         StudyParticipant participant = new StudyParticipant.Builder().withEmail("email@email.com")
                 .withPassword("pAssword1@").withPhone(TestConstants.PHONE).build();
         Validate.entityThrowingException(validator, participant);
     }
     
     @Test
     public void validateTotallyWrongPhone() {
-        validator = new StudyParticipantValidator(study, true);
+        validator = new StudyParticipantValidator(externalIdService, study, true);
         assertValidatorMessage(validator, withPhone("this isn't a phone number", "US"), "phone", "does not appear to be a phone number");
     }
     
+    @Test
+    public void externalIdForNewParticipantValidatedDoesNotExist() {
+        validator = new StudyParticipantValidator(externalIdService, study, true);
+        study.setExternalIdValidationEnabled(true);
+        study.setExternalIdRequiredOnSignup(true);
+        assertValidatorMessage(validator, withExternalId("foo"), "externalId", "is not a valid external ID");
+    }
+    
+    @Test
+    public void externalIdForNewParticipantValidated() {
+        when(externalIdService.getExternalId(study.getStudyIdentifier(), "foo")).thenReturn(EXT_ID);
+        
+        validator = new StudyParticipantValidator(externalIdService, study, true);
+        study.setExternalIdValidationEnabled(true);
+        Validate.entityThrowingException(validator, withExternalId("foo"));
+    }
+    
+    @Test
+    public void externalIdForSavedParticipantOK() {
+        validator = new StudyParticipantValidator(externalIdService, study, false);
+        study.setExternalIdValidationEnabled(true);
+        
+        StudyParticipant participant = new StudyParticipant.Builder().withId("id").withEmail("email@email.com")
+                .withPassword("aAz1%_aAz1%").withExternalId("foo").build();
+        
+        Validate.entityThrowingException(validator, participant);
+    }
+    
+    @Test
+    public void validateExistingManagedExternalId() {
+        when(externalIdService.getExternalId(study.getStudyIdentifier(), "foo")).thenReturn(EXT_ID);
+
+        StudyParticipant participant = new StudyParticipant.Builder().withEmail("email@email.com")
+                .withExternalId("foo").withId("id").withPassword("pAssword1@").build();
+        
+        validator = new StudyParticipantValidator(externalIdService, study, false);
+        study.setExternalIdValidationEnabled(true);
+        Validate.entityThrowingException(validator, participant);
+    }
```

Comment:
Still need a test case for External ID validation required, but not specified.

Reply:
Added (external ID is required but no ID is provided).